Update the specification of the witness process to put the witness proofs in a separate file (#136)
decentralized-identity/trustdidweb: spec/specification.md

in the [DID-to-HTTPS Transformation](#the-did-to-https-transformation) section
of this specification, `did:tdw` and `did:web` DIDs that have the same fully
qualified domain and path transform to the same HTTPS URL, with the exception of
the final file -- `did.json` for `did:web` and `did.jsonl` for `did:tdw`. For
`did:tdw` DIDs using [[ref: witnesses]], another file `witness.json` is also

> find (logically) beside the `did.jsonl` web server file. See the

Reviewer: ```suggestion
found (logically) beside the `did.jsonl` web server file. See the
```

Author: Fixed

6. Generate an HTTPS URL to the expected location of the [[ref: DIDDoc]] by prepending
   `https://`.
7. Append `/did.jsonl` to complete the URL.
   1. If the DID is using [[ref: witnesses]], an extra JSON file containing the
      witness proofs for the [[ref: DID entries]] must be published and
      retrieved during resolution. The URL for the extra file is defined by
      replacing the `/did.jsonl` at the end of the [[ref: DID Log]] URL with
      `/witness.json`.
   2. When this algorithm is used for resolving a DID path (such as
      `<did>/whois` or `<did>/path/to/file` as defined in the section [DID URL
      Handling](#did-url-resolution)), append the DID URL path instead.

> 8. The content type for the `did.jsonl` file **MUST** be `text/jsonl`.

Reviewer: This might be difficult to configure in some cases

Author: Do you have an alternate wording proposal (e.g. change to SHOULD?) or just drop it?

Reviewer: Yes, I think SHOULD is probably strong enough

Author: I set it to SHOULD on this and the one for `did.jsonl`, which I assume has the same issue.

   specification.
4. The JSON object `state` contains the [[ref: DIDDoc]] for this version of the
   DID.

> 5. The JSON array `proof` contains a [[ref: Data Integrity]] proof calculated for

Reviewer 2: sorry.. so pedantic about this phrasing

```suggestion
5. The JSON array `proof` contains a [[ref: Data Integrity]] proof created for
```

   2. `versionTime`
   3. `parameters`
   4. `state` -- the version's [[ref: DIDDoc]].
   5. `proof` -- a [[ref: Data Integrity]] proof for the [[ref: log entry]].

For each entry:

1. Update the currently active [[ref: parameters]] with the [[ref: parameters]] from
   the entry (if any). Continue processing using the now active set of
   [[ref: parameters]].
   - While all [[ref: parameters]] in the first [[ref: Log Entry]] take effect
     immediately, some kinds of [[ref: parameters]] defined in later [[ref:
     entries]] only take effect *after* that entry has been published. For
     example, rotating the authorized keys (in the array `updateKeys`) to update

>      a DID takes effect only *after* the entry in which they are defined has

Reviewer 2: unless we're using prerotation? 😅

Author: Cleaned up to avoid that as an example.  The same example is repeated in a later step in the process. I removed the later step entirely since it is redundant, moving the **MUST** to this step.  The first DRY change!!!

         `nextKeyHashes` list in the [[ref: parameters]].
      6. All other `did:tdw` processing configuration settings as defined by in the
         `parameters` object.
10. If the `parameters` for any of the versions define that some or all of
  the [[ref: DID Log entries]] must be witnessed, further verification of
  the [[ref: witness]] proofs must be carried out, as defined in the [DID
  Witnesses](#did-witnesses) section of this specification.

> 11. If any verifications fail, discard the DID as invalid with an error message.

Reviewer: It's probably acceptable for an individual witness verification to fail as long as the rules are still satisfied?

Author: Good point...I'll adjust.

Author: The point of 11 (where you made the content) is a catch-all for all the verifications (`did:jsonl` file found, SCID calculation, version number is incremented, entryhash validates and so on and so on). I clarified the wording on that.  The "allow witness validations to fail if threshold met" is else where in the spec.

Author: But on that point about failed verifications...

Given that we are using DID:key for the verifications and the proof is across a static object, and was verified before it was inserted into the `did_witness.json` file, I don't think any of them can fail on resolution.  It's not like the DID can disappear like a did:webvh and the key not be found. 
The only reason the proofs in that file would "fail" for a resolver is already covered in the spec -- the witness proof file with new proofs is published before the DID Log with the new version is published.

Reviewer: It does seem unlikely to fail, but I suppose there could be a copy/paste error or it could use an unsupported proof type. The file could even include multiple proof formats for different resolver versions at some point, like in an ECC/post-quantum transition.

Author: OK -- add a tweak (at line 1023 of the specification.md file) that generalizes the point about ignoring proofs that reference a version not in the DID Log 

`Ignore any proofs that do not verify. For example, a `did-witness.json` file  **MAY** contain proofs of pending (unpublished) [[ref: DID Log entries]]. Such proofs **MUST** be ignored by resolvers.`

  collaboration with [[ref: witnesses]] prior to publication. For details of
  this data and its usage in the approvals process, see the
  [DID Witnesses](#did-witnesses) section of this specification.
  - A `witness` property in the first [[ref: DID log entry]] is immediately
    "active" and used to define the [[ref: witnesses]] and necessary [[ref: threshold]]
    for witnessing the initial [[ref: log entry]]. In all other [[ref: DID log entries]],
    a `witness` property becomes active **after** the publication of its entry

>     -- meaning its [[ref: log entry]] **MUST** be witnessed by active

Reviewer 2: I don't think we say anything about whether `witness` can be disabled after it's been enabled or is this forbidden?

Author: To be added as an issue.

of the DID are valid and approved.

##### Witnessing a DID Version Update

The following process is used to witness a DID version update:

- The [[ref: DID Controller]] prepares the full [[ref: DID Log Entry]] (including the
  `proof` element) for the new version of DID, and shares it with the active [[ref: witnesses]].
  - The specification leaves to implementers *how* the [[ref: log entry]] data is provided to the [[ref: witnesses]].
- The [[ref: witnesses]] ***MUST** hold their own copy of the published [[ref:
  DID Log]] prior to the version being witnessed.
- Each [[ref: witness]] verifies the [[ref: DID Log Entry]], as defined by this
  specification. If not, the [[ref: witnesses]] **MUST NOT** approve the [[ref: log entry]].
- Each [[ref: witness]] determines (based on the governance of the ecosystem)
  if they approve of the DID version update.
  - The meaning of "approve" for any given implementation is outside the scope of this specification.
- If the verification is successful and approval granted, the [[ref: witness]]
  sends a [[ref: Data Integrity]] proof across the [[ref: DID log entry]] (minus

>   the `proof` item) to the [[ref: DID Controller]] signed by the [[ref: witness]]'s key.

Reviewer: Maybe "following the procedure in §... for signing log entries"?

Author: Added -- hopefully the wording is sufficient.

    earlier [[ref: entries]] to reduce the size of the log. Only the set of approval
    proofs on the last [[ref: log entry]] are needed because of the chaining of the
    proofs via the use of the `entryHash` challenge.
  - Removing the prior entry [[ref: witness]] proofs does not affect the verifiability of
    the DID because the `entryHash` calculation does not include the `proof` property.
  - The specification leaves to implementers how the proofs are conveyed to
    the [[ref: DID Controller]].

As with the handling of the `updateKeys`, [[ref: DID Log Entry]] changes require
proofs from the the [[ref: witnesses]] active *prior* to the publication of a
new version. If a new version changes the list of [[ref: witnesses]], that
change must be approved by the *prior* [[ref: witnesses]]. For the first entry
in the [[ref: DID Log]], the [[ref: witnesses]] listed in that entry must
approve the version, since there are no "prior" [[ref: witnesses]].

The data model for the `witness` [[ref: parameter]] is as follows.
The threshold design borrows from the [[ref: verifiable conditions]]
specification.

```json
"witness" : {

>   "threshold": n,

Reviewer 2: Is this defined elsewhere now?

>             "version_num": "1",

Author: Sounds good.  I think I would prefer to use this data structure vs. the DI Array.  I think it is a bit subtle how things link together.  I think we can merge this (after the change to version_id) and then debate the proof array idea.

>             "witness": "did:key:z82LkvR3CBNkb...",

Reviewer 2: This property isn't technically required as the `proof` will contain the `verificationMethod` property

Reviewer 2: I'm fine with it in there -- makes for easy comparisons to the parameter entry

       },
       ...
   ]
}
```

Where:

- `version_num` is the prefix (prior to the literal `-`) of the `version_id` of the [[ref: DID log entry]] to which the proof applies.
- `witness` is the DID of the witness.
- `proof` is the proof of the [[ref: DID Log Entry]], excluding the `proof` item, identified by the `version_num`.

Only the two most recent proofs from each [[ref: witness]] are retained in the
`did-witness.json` file. When a new, verified proof from a [[ref: witness]] is received,
the [[ref: DID Controller]] adds it to the file. If two other proofs from that

> [[ref: witness]] are found in the file, the oldest **MUST** be removed. The presence of a valid

Author: No -- I think **SHOULD** is good enough -- with the addition of the purpose is to keep the file size down.

> [[ref: witness]] are found in the file, the oldest **MUST** be removed. The presence of a valid

Reviewer 2: ```suggestion
[[ref: witness]] are found in the file, the oldest **SHOULD** be removed. The presence of a valid
```

> [[ref: witness]] are found in the file, the oldest **MUST** be removed. The presence of a valid

Author: Fixed

proof is an attestation from that [[ref: witness]] that the current and all prior versions
of the DID are valid and approved.

##### Witnessing a DID Version Update

The following process is used to witness a DID version update:

- The [[ref: DID Controller]] prepares the full [[ref: DID Log Entry]] (including the
  `proof` element) for the new version of DID, and shares it with the active [[ref: witnesses]].
  - The specification leaves to implementers *how* the [[ref: log entry]] data is provided to the [[ref: witnesses]].
- The [[ref: witnesses]] ***MUST** hold their own copy of the published [[ref:
  DID Log]] prior to the version being witnessed.
- Each [[ref: witness]] verifies the [[ref: DID Log Entry]], as defined by this
  specification. If not, the [[ref: witnesses]] **MUST NOT** approve the [[ref: log entry]].
- Each [[ref: witness]] determines (based on the governance of the ecosystem)
  if they approve of the DID version update.
  - The meaning of "approve" for any given implementation is outside the scope of this specification.
- If the verification is successful and approval granted, the [[ref: witness]]

>   sends a [[ref: Data Integrity]] proof across the [[ref: DID log entry]] (minus

Author: Fixed -- the first change was  "creates and sends..."

  the `proof` item) to the [[ref: DID Controller]] signed by the [[ref: witness]]'s key.
  - The proof generation process of the witness **MUST** match the one used by

>    the DID Controller in generating its proof across the [[ref: DID log entry]],

Reviewer 2: ```suggestion
   the DID Controller in generating its proof for the [[ref: DID log entry]],
```

Author: Fixed and other references to "across" when talking about a proof were changed as well.  I left "across" when talking about a hash, but perhaps that is also not right?

   as defined in the [Authorized Keys](#authorized-keys) section of this
   specification. 
- The DID Controller updates the [[ref: witnesses]]'s proofs in the
  `did-witness.json` file with the received proofs, and publishes the updated file.
  - The specification leaves to implementers how [[ref: witness]] proofs are
    conveyed to the [[ref: DID Controller]].
  - The [[ref: DID Controller]] **MAY** publish the updated `did-witness.json` file
    as new witness proofs are added to the file.

>   - The [[ref: DID Controller]] **MUST** publish the update `did-witness.json` file

Reviewer 2: ```suggestion
  - The [[ref: DID Controller]] **MUST** publish the updated `did-witness.json` file
```

Author: Fixed

    **after** a threshold of witness proofs have been received and **before** the
    witnessed [[ref: DID Log]] file is published.

##### Verifying Witness Proofs During Resolution

A `did:tdw` resolver **MUST** verify that all [[ref: DID Log entries]] that have
active [[ref: witnesses]] have a threshold of active witnesses approving the [[ref: log entry]].
To do so, resolvers must:

- Successfully complete the non-[[ref: witness]] verifications of the [[ref: DID Log]].
- Verify the [[ref: witness]] proofs that reference [[ref: DID Log entries]] in the [[ref: DID Log]].
  - Ignore any witness proofs that do not verify. For example, a `did-witness.json` file
    **MAY** contain proofs of pending (unpublished) [[ref: DID Log entries]].
    Such proofs **MUST** be ignored by resolvers.
- For each [[ref: DID log entry]] requiring witnessing, the resolver **MUST**

>   verify that the `did-witness.json` file contains verified [[ref: witness]] proofs from a

Reviewer 2: ```suggestion
  verify that the `did-witness.json` file contains verified [[ref: witness]] [[ref: Data Integrity]] proofs from a
```

>   threshold of active [[ref: witnesses]] across the current or **later** log entries. If

Reviewer 2: ```suggestion
  threshold of active [[ref: witnesses]] of the current or **later** log entries. If
```

#### DID Witnesses

The [[ref: witness]] process for a DID provides a way for collaborators to work
with the [[ref: DID Controller]] to "witness" the publication of new versions of
the DID. This specification defines the technical mechanism for using [[ref:
witnesses]]. Governance and policy questions about when and how to use the
technical mechanism are outside the scope of this specification.

> Witnesses can prevent a [[ref: DID Controller]] from updating/removing DID

Reviewer 2: ```suggestion
Witnesses can prevent a [[ref: DID Controller]] from updating/removing
```

Author: Fixed

versions of a DID without detection by the witnesses. [[ref: Witnesses]] are
also a further mitigation against malicious actors compromising both a [[ref:
DID Controller]]'s authorization key(s) to update the DID, and the [[ref: DID
Controller]]'s web site where the [[ref: DID log]] is published. With both
compromises, a malicious actor could take control over the DID by rewriting the

> [[ref: DID Log]] using the keys they have comprised. By adding [[ref:

Reviewer 2: ```suggestion
[[ref: DID Log]] using the keys they have compromised. By adding [[ref:
```

Author: Fixed

witnesses]] to monitor and approve each version update, a malicious actor cannot
rewrite the previous history without having compromised a sufficient number of
[[ref: witnesses]], the [[ref: DID Controller]]'s key(s), and the Web Server on
which the [[ref: DID Log]] is published.

##### Witness Lists

The list of DIDs that witness DID updates are defined in the `witness`
parameter, as described in the [Parameters](#didtdw-did-method-parameters)
section of this specification. Once the first `witness` parameter has been added
to a version, there is always an active list of witnesses, and a [[ref: threshold]] of
the active witnesses must provide verified proofs about an update before the

> update can be published. If the a DID version contains a new (replacement) list

Reviewer 2: ```suggestion
update can be published. If a DID version contains a new (replacement) list
```

Author: Fixed

of witnesses (by including a new `witness` [[ref: parameter]]) that new list
becomes active **AFTER** the new version is published.

##### Witness DIDs and Reputation

`did:tdw` witness DIDs **MUST** be `did:key` DIDs.

If there is a need in an ecosystem to identify who the witnesses are, a
mechanism should be defined by the governance of the ecosystem, such as the
entry of the DID in a [[ref: Trust Registry]]. Such mechanisms are outside the
scope of this specification.

##### The `witness` Parameter

The `witness` element in a [[ref: parameters]] object of a [[ref: DID Log
entry]] has the following data structure:

```json

"witness" : {
  "threshold": n,
  "witnesses" : [
      {
         "id": "<did:key DID of witness>",
         "weight": n
      }
   ]
}

```

where:

> - threshold: an integer that must be attained or surpassed by the sum of the witnesses weights for a DID log entry to be considered approved.
> - witnesses: an array of witnesses, each including the required fields:
>   - id: the DID of the witness. The DID **MUST** be a `did:key` DID.
>   - weight: an integer that is the weight given to this witness's approval
> 
> ##### Witness Threshold Algorithm

Reviewer 2: ```suggestion
- `threshold`: an integer that must be attained or surpassed by the sum of the witnesses weights for a DID log entry to be considered approved.
- `witnesses`: an array of witnesses, each including the required fields:
  - `id`: the DID of the witness. The DID **MUST** be a `did:key` DID.
  - `weight`: an integer that is the weight given to this witness's approval

##### Witness Threshold Algorithm
```

The use of the [[ref: threshold]] and weighted approvals (versus needing
approvals from all [[ref: witnesses]]) is to prevent faulty [[ref: witnesses]]
from blocking the publishing of a new version of the DID. To determine if the
[[ref: threshold]] has been met, sum the `weight` integer of the received

> approvals and if it equal to or more than the `threshold`, the update can be

Reviewer 2: ```suggestion
approvals and if it is greater than or equal to the `threshold`, the update can be
```

[[ref: DID Log]].

For example, if there are three [[ref: witnesses]] with a `weight` of `1`, a
fourth with a weight of `2`, and a `threshold` of `3`, the [[ref: threshold]] is met
by either the fourth plus any one of the other [[ref: witnesses]] (`2+1`), or
all of the first three witness (`1+1+1`) providing an approving `proof`.

##### The Witness Proofs File

Proofs from witnesses are retained in a separate file (`did-witness.json`) from the
[[ref: DID Log]]. The same [DID to HTTPS
Transformation](#the-did-to-https-transformation) used for the [[ref: DID Log]]
is used to locate the `did-witness.json` resource, with only the last element
changed (`did.jsonl` to `did-witness.json`). The media type of the file **SHOULD** be
`text/json`.

The data model for the `did-witness.json` file is:

```json
{
    [
       {
            "versionId": "1-QmfGEUAcMpzo25kF2Rhn8L5FAXysfGnkzjwdKoNPi615XQ",
            "witness": "did:key:z82LkvR3CBNkb...",
            "proof": { ... }
       },
       ...
   ]
}
```

Where:

- `versionId` is the `versionId` of the [[ref: DID log entry]] to which the proof applies.
- `witness` is the DID of the witness.

> - `proof` is the proof of the [[ref: DID Log Entry]] (excluding the `proof` item) identified by the `version_id`.

Reviewer 2: ```suggestion
- `proof` is the proof of the [[ref: DID Log Entry]] (excluding the `proof` item) identified by the `versionId`.
```

Author: Fixed

To limit the size of the file, only the two most recent proofs from each [[ref: witness]] are retained in the

> `did-witness.json` file. When a new, verified proof from a [[ref: witness]] is received,

Reviewer 3: Does it make sense to instead only keep the proofs about the two most recent log update entries?

Reviewer: The resolver may still need to check proofs from witnesses that aren't active anymore, and aren't signing the latest log entries.

Author: I'll add this to an issue and we can debate it.

Reviewer: We could be a bit more precise and say, for each witness:
- Keep the latest proof from among the previous versions
- Add the proof against the unpublished version, if there is one

This would mean if they didn't sign the latest transaction, there would only be one proof for that witness.

Author: Let’s also do this as an add on. I like it.  We might get a little more thorough and say that applies to all witnesses, so any removed witness never has more than one proof.  Gets a little nuanced. I’ll add an issue.

the [[ref: DID Controller]] adds it to the file. If two other proofs from that
[[ref: witness]] are found in the file, the oldest **SHOULD** be removed. The presence of a valid
proof is an attestation from a [[ref: witness]] that the current **and all prior** versions
of the DID have been verified and approved by that witness.

##### Witnessing a DID Version Update

The following process is used to witness a DID version update:

- The [[ref: DID Controller]] prepares the full [[ref: DID Log Entry]] (including the

>   `proof` element) for the new version of DID, and shares it with the active [[ref: witnesses]].

Reviewer 2: ```suggestion
  `proof` element) for the new version of the DID, and shares it with the active [[ref: witnesses]].
```

Author: Fixed

  - The specification leaves to implementers *how* the [[ref: log entry]] data is provided to the [[ref: witnesses]].
- The [[ref: witnesses]] ***MUST** hold their own copy of the published [[ref:
  DID Log]] prior to the version being witnessed.
- Each [[ref: witness]] verifies the [[ref: DID Log Entry]], as defined by this

>   specification. If not, the [[ref: witnesses]] **MUST NOT** approve the [[ref: log entry]].

Reviewer 2: ```suggestion
  specification. If not verified, the [[ref: witnesses]] **MUST NOT** approve the [[ref: log entry]].
```

Author: Fixed

- Each [[ref: witness]] determines (based on the governance of the ecosystem)
  if they approve of the DID version update.
  - The meaning of "approve" for any given implementation is outside the scope of this specification.
- If the verification is successful and approval granted, the [[ref: witness]]

>   creates and sends a [[ref: Data Integrity]] proof for the [[ref: DID log entry]] (minus

Reviewer 3: Wouldn't just signing the versionId be sufficient since the versionId is a sort of hash of the entry? Meaning if the entry validation passes, a proof over the versionId only would be sufficient?

Reviewer 2: hmm interesting.. I don't hate this idea. For implementation I was originally thinking I would just verify a proof set all at once so this would make it two separate verifications but I don't think that's a big deal.

Reviewer 3: that's an interesting point, so you just match the elements based on the version Id and merge the proofs into 1 array

Reviewer 3: could we then just have an object of key/value where the key is the versionId and the value is a proof set?
```
{
     "version-1": [{ witness-1-proof },{ witness-2-proof },{ witness-3-proof }],
     "version-2": [{ witness-1-proof },{ witness-2-proof },{ witness-3-proof }],
     "version-3": [{ witness-1-proof },{ witness-2-proof },{ witness-3-proof }],
}
```

Reviewer 3: my idea was having theses objects with a normal data integrity signature applied to them, not the whole entry
```
[
     {
          "versionId": "version-1",
          "proof": [{ witness-1-proof },{ witness-2-proof },{ witness-3-proof }]
      },
     {
          "versionId": "version-2",
          "proof": [{ witness-1-proof },{ witness-2-proof },{ witness-3-proof }]
      },
]

```

Reviewer: Yes, that might be a good simplification. If we're using DI then it still needs a JSON object payload though, so possibly something like `{"versionId":...}`?

Reviewer 3: witness proof for the same entry/update should definitely be grouped together as a set

Reviewer 3: Yea, because all the witness really need to do is say "yes, I verified/validated this entry and all is good". The `versionId` is derived from the entry itself and is part of the verification, so it would carry the same idea IMHO.

Author: I'll add to the issue and we can debate it.

  - The specification leaves to implementers how [[ref: witness]] proofs are
    conveyed to the [[ref: DID Controller]].
  - The [[ref: DID Controller]] **MAY** publish the updated `did-witness.json` file
    as new witness proofs are added to the file.
  - The [[ref: DID Controller]] **MUST** publish the updated `did-witness.json` file
    **after** a [[ref: threshold]] of witness proofs have been received and **before** the
    witnessed [[ref: DID Log]] file is published.

##### Verifying Witness Proofs During Resolution

A `did:tdw` resolver **MUST** verify that all [[ref: DID Log entries]] that have
active [[ref: witnesses]] have a [[ref: threshold]] of active witnesses approving the [[ref: log entry]].
To do so, resolvers must:

- Successfully complete the non-[[ref: witness]] verifications of the [[ref: DID Log]].
- Verify the [[ref: witness]] proofs that reference [[ref: DID Log entries]] in the [[ref: DID Log]].
  - Ignore any witness proofs that do not verify. For example, a `did-witness.json` file
    **MAY** contain proofs of pending (unpublished) [[ref: DID Log entries]].
    Such proofs **MUST** be ignored by resolvers.

> - For each [[ref: DID log entry]] requiring witnessing, the resolver **MUST**

Reviewer 2: An entry in the past might have required witnessing at the time but those proofs have been discarded. This reads as though they need to verify those old proofs. Later in the paragraph it says "current or later" so I think it's saying the right things here it's just a bit misleading. Maybe we move the "current or later" text earlier in the paragraph?

Author: I tweaked the wording a little, but I think it is "clear enough" (although it is a bit tricky).  We can work on it further in later PRs.  I don't think it is wrong as it is.

entry of the DID in a [[ref: Trust Registry]]. Such mechanisms are outside the
scope of this specification.

##### The `witness` Parameter

The `witness` element in a [[ref: parameters]] object of a [[ref: DID Log
entry]] has the following data structure:

```json

"witness" : {
  "threshold": n,

>   "selfWeight": n,

Reviewer 2: I think we've talked about this before... but do we really need `selfWeight`? They will always be signing their own entry.. It feels like added complexity for little gain as they could just decrease the threshold and not include their own proof in that calculation

{
    [
       {

>             "version_id": "1-QmfGEUAcMpzo25kF2Rhn8L5FAXysfGnkzjwdKoNPi615XQ",

Reviewer 2: to match the rest of the spec

```suggestion
            "versionId": "1-QmfGEUAcMpzo25kF2Rhn8L5FAXysfGnkzjwdKoNPi615XQ",
```

Author: Doh... thanks!!!

Where:

> - `version_id` is the `version_id` of the [[ref: DID log entry]] to which the proof applies.

Reviewer 2: ```suggestion
- `versionId` is the `versionId` of the [[ref: DID log entry]] to which the proof applies.
```

Author: Fixed
